### PR TITLE
New Resource: aws_dx_lag

### PR DESCRIPTION
**Pull request by atsushi-ishibashi**
```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsDxLag_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsDxLag_basic -timeout 120m
=== RUN   TestAccAwsDxLag_basic
--- PASS: TestAccAwsDxLag_basic (40.91s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	40.984s
```

### CODE REVIEW

**Comment by atsushi-ishibashi** [comment]
Related #876

File: aws/resource_aws_dx_lag.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		LagId: aws.String(lagId),
+	}
+	resp, err := conn.DescribeLags(input)
+	if err != nil {
+		d.SetId("")
+		return err
+	}
+	if len(resp.Lags) != 1 {
+		d.SetId("")
+		return fmt.Errorf("[ERROR] Number of DX Lag (%s) isn't one, got %d", lagId, len(resp.Lags))
```
Comment (radeksimko) [inline aws/resource_aws_dx_lag.go]: I'm ok with erroring out if we get != 1 lags, but I don't think it's a reason for removing the LAG from state. We should only be removing it from state if it's not found - i.e. `len(resp.Lags) < 1`

File: aws/resource_aws_dx_lag.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+func resourceAwsDxLagRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dxconn
+
+	lagId := d.Id()
+	input := &directconnect.DescribeLagsInput{
+		LagId: aws.String(lagId),
+	}
+	resp, err := conn.DescribeLags(input)
+	if err != nil {
+		d.SetId("")
```
Comment (radeksimko) [inline aws/resource_aws_dx_lag.go]: I can't find any particular error that would mean LAG is gone in this context: http://docs.aws.amazon.com/directconnect/latest/APIReference/API_DescribeLags.html#c
so I don't think error here is a reason for removing the LAG from state. 🤔

File: aws/resource_aws_dx_lag.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		return err
+	}
+	if len(resp.Lags) != 1 {
+		d.SetId("")
+		return fmt.Errorf("[ERROR] Number of DX Lag (%s) isn't one, got %d", lagId, len(resp.Lags))
+	}
+	if d.Id() != *resp.Lags[0].LagId {
+		d.SetId("")
+		return fmt.Errorf("[ERROR] DX Lag (%s) not found", lagId)
```
Comment (radeksimko) [inline aws/resource_aws_dx_lag.go]: It makes sense to error out if wrong ID was found (not sure how/why would that happen, but strange things do happen sometimes), but I'm not sure if it's the reason for removing LAG from state.

File: aws/resource_aws_dx_lag.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	input := &directconnect.DeleteLagInput{
+		LagId: aws.String(d.Id()),
+	}
+	_, err := conn.DeleteLag(input)
+	if err != nil {
+		return err
+	}
+	deleteStateConf := &resource.StateChangeConf{
+		Pending:    []string{directconnect.LagStateAvailable, directconnect.LagStateRequested, directconnect.LagStatePending},
+		Target:     []string{directconnect.LagStateDeleted, directconnect.LagStateDeleting},
```
Comment (radeksimko) [inline aws/resource_aws_dx_lag.go]: Is there any reason we should treat `directconnect.LagStateDeleting` as `Target` state here?

File: aws/resource_aws_dx_lag.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		Read:   resourceAwsDxLagRead,
+		Update: resourceAwsDxLagUpdate,
+		Delete: resourceAwsDxLagDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"band_width": &schema.Schema{
```
Comment (radeksimko) [inline aws/resource_aws_dx_lag.go]: Nitpick, but is there any reason to drift away from the name in the API? i.e. can't we just call it `connections_bandwidth`?

File: aws/resource_aws_dx_lag.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateDxLagBandWidth,
+			},
+			"location": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"num_connections": &schema.Schema{
```
Comment (radeksimko) [inline aws/resource_aws_dx_lag.go]: Nitpick, but is there any reason to drift away from the name in the API? i.e. can't we just call it `number_of_connections `?

File: aws/resource_aws_dx_lag_test.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		return nil
+	}
+}
+
+func testAccDxLagConfig(rName string) string {
+	return fmt.Sprintf(`
+    resource "aws_dx_lag" "hoge" {
+      name = "tf-dx-lag-%s"
+      band_width = "1Gbps"
+      location = "EqDC2"
```
Comment (radeksimko) [inline aws/resource_aws_dx_lag_test.go]: Do you mind changing the location to a one which is available in `us-west-2` as all of our tests run in that region? e.g. `EqSe2`

File: aws/resource_aws_dx_lag.go (hunk 5 above)
Comment (atsushi-ishibashi) [reply aws/resource_aws_dx_lag.go]: I totally agree with your reviews!

File: aws/resource_aws_dx_lag.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	input := &directconnect.DescribeLagsInput{
+		LagId: aws.String(lagId),
+	}
+	resp, err := conn.DescribeLags(input)
+	if err != nil {
+		return err
+	}
+	if len(resp.Lags) != 1 {
+		if len(resp.Lags) < 1 {
+			d.SetId("")
```
Comment (radeksimko) [inline aws/resource_aws_dx_lag.go]: I think in this case (when Lag was removed) we don't want to error out - at least it's a convention not to do that.

Also - which is rather a nitpick - we can move this condition above or below the other condition and avoid the nesting & indentation so it's easier to read.

Comment (atsushi-ishibashi) [reply aws/resource_aws_dx_lag.go]: I understood the convention 🙇

**Pull request merged**